File: gachisa-backend/src/main/java/com/gachisa/payment/service/PaymentRecoveryService.java
package com.gachisa.payment.service;

import com.gachisa.global.exception.CustomException;
import com.gachisa.payment.client.PgClient;
import com.gachisa.payment.client.dto.PgPaymentQueryResult;
import com.gachisa.payment.dto.PaymentResponse;
import com.gachisa.payment.service.dto.RecoveryPreparation;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class PaymentRecoveryService {

    private final PaymentRecoveryStateService recoveryStateService;
    private final PgClient pgClient;

    public PaymentResponse recover(Long paymentAttemptId) {
        RecoveryPreparation preparation = recoveryStateService.prepare(paymentAttemptId);
        if (!preparation.queryRequired()) {
            return preparation.existingResponse();
        }

        try {
            PgPaymentQueryResult result = pgClient.getPayment(preparation.paymentKey());
            return recoveryStateService.apply(paymentAttemptId, result);
        } catch (CustomException exception) {
            recoveryStateService.recordFailure(paymentAttemptId, exception.getErrorCode());
            throw exception;
        }
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/service/PaymentRecoveryService.kt
package com.gachisa.payment.service

import com.gachisa.global.exception.CustomException
import com.gachisa.payment.client.PgClient
import com.gachisa.payment.dto.PaymentResponse
import com.gachisa.payment.metric.PaymentMetrics
import org.springframework.stereotype.Service

@Service
class PaymentRecoveryService(
    private val recoveryStateService: PaymentRecoveryStateService,
    private val pgClient: PgClient,
    private val paymentMetrics: PaymentMetrics,
) {
    fun recover(paymentAttemptId: Long): PaymentResponse {
        val preparation = recoveryStateService.prepare(paymentAttemptId)
        if (!preparation.queryRequired) {
            paymentMetrics.recordPaymentRecovery("skipped")
            return preparation.existingResponse!!
        }

        try {
            val result = pgClient.getPayment(preparation.paymentKey)
            val response = recoveryStateService.apply(paymentAttemptId, result)
            paymentMetrics.recordPaymentRecovery("success")
            return response
        } catch (exception: CustomException) {
            recoveryStateService.recordFailure(paymentAttemptId, exception.getErrorCode())
            paymentMetrics.recordPaymentRecovery("failed")
            throw exception
        }
    }
}
